Support for gradually ramping fan speeds

bsquizz opened the pull request:
The goal of this PR is to allow fan speeds to ramp up and down more smoothly. With changing temperatures you'll often hear the fan speed "jump" every few seconds.  It adds a "FanSpeedController" which runs in a background thread and handles gradually changing the fan speeds.

Each config strategy can now have a 'ramp up time' and a 'ramp down time'. If the fan speed is told to increase/decrease, it will do so gradually over a 'ramp up time'/'ramp down time'.

There are 2 exceptions to this:
* If a configurable 'ramp quickly temperature' is hit, the fan speed will begin to ramp at a faster pace.
* If a configurable 'critical temperature' is hit, the moving average ~~calculation is thrown out~~ is weighted so that the fan speed immediately begins to ramp up.

The FanSpeedController handles the ramp up / down by looking at the "current speed" vs "desired speed" and figuring out how to gradually get to the desired speed. For example: if current speed is '3000', desired speed is '4000', and 'ramp_up_time' is '3 seconds', the fan speed will make small increases of ~67 rpm every 0.2 seconds -- which would be 15 times over 3 seconds.

The current 'step' time is set to 0.2 and it is a totally arbitrary number I picked. This means 'ectool' is run every 0.2 seconds during a ramp up/ramp down. I am not sure if there is a negative impact to running ectool so frequently... so far I haven't noticed a major problem with doing this.

bsquizz committed:
Change speeds with smooth ramping

bsquizz committed:
Change speeds with smooth ramping

bsquizz committed:
Remove unused 'i'

bsquizz committed:
Remove unused 'i'

bsquizz committed:
More tweaking

bsquizz committed:
More tweaking

bsquizz committed:
Add config options, quick ramp/critical temp (assisted-by: Cursor+claude-4-sonnet)

bsquizz committed:
Add config options, quick ramp/critical temp (assisted-by: Cursor+claude-4-sonnet)

bsquizz committed:
Weight the moving average when critical temp is hit

bsquizz committed:
Weight the moving average when critical temp is hit

bsquizz committed:
Use 100 degrees C as critical temp

bsquizz committed:
Use 100 degrees C as critical temp

leopoldhub commented:
Hi, tank you for your contribution !

I sadly will not have time to check this PR in details for the next week, but will gladly take a look after.

Sorry for the delay.

bsquizz committed:
Use threading.Event, fix miscalcs due to int rounding

bsquizz committed:
Use threading.Event, fix miscalcs due to int rounding

nyabinary commented:
> Hi, tank you for your contribution !
> 
> I sadly will not have time to check this PR in details for the next week, but will gladly take a look after.
> 
> Sorry for the delay.

Any updates/

leopoldhub commented:
Hi, sorry for the... very-much-longer-than-a-week delay

After more reflection, I believe that those options are add too much complexity to the configuration, and makes the temperature curve difficult to represent/visualize

I understand your need, but will not accept this PR before we find a solution to this issue

bsquizz commented:
I could hard code the values for the ramp up / ramp down instead of complicating the configuration. Is that what you are looking for?

leopoldhub commented:
Hi, 

Not at all, I would like to avoid hard coded values whenever possible

Here is additional feedback

Configuration should:

1. Be small and easy to read
2. Be intuitive
3. Be predictable

In your proposed changes, you inadvertently add complexity to those three:

1. 5 new parameters per profile
2. Names (or what they represent and how to interpret them)
are not easy to understand if you do not have the code or the doc on the side
4. The curve is now more complex to imagine/visualize, and less predictable

On an other note, I would prefer fan speed calculation to be done at the `fanSpeedUpdateFrequency` frequency, and be as close as possible to a mathematical formula instead of a program with external additional clock/frequency handling.

That said, I do believe that you have a point about fan speed jump, and what you built is a great start to find a solution.

I do not have a solution at hand, but I would expect the following:

1. 2 or less new options per profile
We can also imagine having additional properties on the curve point themselves
2. Simple names that are easy to represent and understand
3. Calculation must be done synchronously in adapt_speed

I am open to discussion
If you have any question, feel free to ask them too!

bsquizz committed:
Change defaults

bsquizz commented:
I can improve the config.

>  I would prefer fan speed calculation to be done at the fanSpeedUpdateFrequency frequency

The reason it is done async is because there's now 2 things to consider:
1. the target fan speed you want to be at
2. how you're going to smoothly change from your current speed to the target

Let's say the speed update frequency is set to 5 seconds. And also let's say in that time, the temperature goes from 50 to 90. I don't want the fan to immediately jump directly from 0 rpm to 7000rpm, that is exactly the behavior I'm trying to fix here. Something needs to change its speed async so it gradually goes from 0 rpm to 7000 rpm over the course of 1 second or 2 seconds.

bsquizz committed:
Simplify fan speed ramping to single rampDuration config option

Replace 5 config options (rampUpTime, rampDownTime, rampQuicklyTemperature,
rampQuicklyTime, criticalTemperature) with a single optional rampDuration
option that controls how long fan speed transitions take (defaults to 2s).

FanSpeedController runs in a background thread and smoothly interpolates
fan RPM from current to target speed over rampDuration using 0.2s steps.
If a new target arrives mid-ramp, it interrupts and starts a new ramp.

Also refactors adapt_speed by extracting get_target_speed helper.

No changes to default config values or existing strategy parameters.

(assisted-by: opencode+claude)

bsquizz commented:
Here's where we are at with the latest commit:

- Kept the async FanSpeedController for the reasons listed above -- I am not yet sure of a simpler way to smooth out the speed change between 'current' and 'target'.

- Cleaned up the config (removed the complex config that had 5 different options), only 1 new optional config option: `rampDuration`
  - Controls the time (in seconds) for the fan RPM to ramp up or down to the desired target speed
  - Defaults to 2 seconds if omitted
  - No existing configs need to change

- Removed critical temperature "weighted-average injection" logic to reduce complexity

bsquizz committed:
Update README.md

leopoldhub commented:
Hey!

Nice changes, I find it way more intuitive now!

I still have some feedback:

The fan speed can never be updated quicker than the value defined in `fanSpeedUpdateFrequency`, as it is the main execution clock

Considering that, would it be possible to process the ramp effect once every clock action in `get_target_speed`, and use the present `fanSpeedUpdateFrequency` value as a reference to know how many steps to execute?

If no ramp speed is configured, the default behaviour is have the target speed be the desired speed
